rust/src/pool/mod.rs

```diff
@@ -0,0 +1,409 @@
+use std::{
+    ops::DerefMut,
+    sync::{Arc, LockResult, RwLock},
+};
+
+use crate::error::Error;
+
+/// An abstraction for a pool of thread-safe elements
+pub trait Pool {
+    /// The key type used to identify entries in the pool.
+    type Key;
+    /// The type of elements indexed by the pool.
+    type Type;
+
+    /// Retrieves an entry from the pool.
+    /// The concrete type returned by the get operation may not be the concrete type `Self::Type`
+    /// but instead a wrapper type which behaves like `Self::Type`. This abstraction allows for
+    /// the pool to store complex types which track the lifetime of stored objects, enabling
+    /// nested catching.
+    fn get(
+        &self,
+        id: Self::Key,
+    ) -> Result<PoolEntry<impl DerefMut<Target = Self::Type> + Send + Sync + 'static>, Error>;
+
+    /// Stores the value in the pool and reserves an ID for it.
+    fn set(&self, value: Self::Type) -> Result<Self::Key, Error>;
+
+    /// Deletes the entry with the given ID from the pool
+    /// The ID may be reused in the future, when creating a new entry by calling [`Pool::set`].
+    fn delete(&self, id: Self::Key) -> Result<(), Error>;
+
+    /// Flushes all pending operations to the underlying IO hierarchy.
+    #[allow(dead_code)]
+    fn flush(&self) -> Result<(), Error>;
+}
+
+/// An element retrieved from the pool which can be locked before reading to enable safe concurrent
+/// access. Instances of this object shall not be kept for longer than the scope where they are
+/// needed, failing to release these nodes will lead to resources leackage.
```

Comment:
```suggestion
/// needed, failing to release these nodes will lead to resources leakage.
```

Whether or not resources will be leaked depends on the pool implementation (it might not hold on to a copy of the Arc internally), so I don't think it makes sense to prescribe the behavior on this type. I think we're being a bit too paranoid about the whole "leaking nodes" scenario.

Reply:
True, especially now that the lifetime of the returned type is `static` the implementer can move items out of the pool.
Removing the last sentence